src/plan.ts

```diff
@@ -0,0 +1,186 @@
+  const charmcraftFiles = await (
+    await glob.create(path.join(workingDir, '**', 'charmcraft.yaml'))
+  ).glob()
+  return charmcraftFiles.map((charmcraftFile: string) => {
+    const file = path.join(
+      workingDir,
+      path.relative(workingDir, charmcraftFile)
+    )
+    const charmcraft = yaml.load(
+      fs.readFileSync(charmcraftFile, { encoding: 'utf-8' })
+    ) as object
+    // @ts-ignore
+    let name: string
+    if ('name' in charmcraft) {
+      name = charmcraft['name'] as string
+    } else {
+      const metadataFile = path.join(
+        path.dirname(charmcraftFile),
+        'metadata.yaml'
+      )
+      const metadata = yaml.load(
+        fs.readFileSync(metadataFile, { encoding: 'utf-8' })
+      ) as object
+      if (!('name' in metadata)) {
+        throw new Error(`unknown charm name (${workingDir})`)
+      }
+      name = metadata['name'] as string
+    }
+    return {
+      type: 'charm',
+      name,
+      source_file: file,
+      source_directory: path.dirname(file),
+      output_type: 'file',
+      output: sanitizeArtifactName(`${id}__build__output__charm__${name}`)
+    }
+  })
+}
+
+async function planBuildRock(
+  workingDir: string,
+  id: string,
+  outputType: 'file' | 'registry'
+): Promise<BuildPlan[]> {
+  const rockcraftFiles = await (
+    await glob.create(path.join(workingDir, '**', 'rockcraft.yaml'))
+  ).glob()
+  return rockcraftFiles.map((rockcraftFile: string) => {
+    const file = path.join(workingDir, path.relative(workingDir, rockcraftFile))
+    const rockcraft = yaml.load(
+      fs.readFileSync(rockcraftFile, { encoding: 'utf-8' })
+    )
+    // @ts-ignore
+    const name = rockcraft['name']
+    return {
+      type: 'rock',
+      name,
+      source_file: file,
+      source_directory: path.dirname(file),
+      output_type: outputType,
+      output: sanitizeArtifactName(`${id}__build__output__rock__${name}`)
+    }
+  })
+}
+
+async function planBuildDockerImage(
+  workingDir: string,
+  id: string,
+  outputType: 'file' | 'registry'
+): Promise<BuildPlan[]> {
+  const dockerFiles = await (
+    await glob.create(path.join(workingDir, '**', '*.Dockerfile'))
+  ).glob()
+  return dockerFiles.map((dockerFile: string) => {
+    const file = path.join(workingDir, path.relative(workingDir, dockerFile))
+    const name = path.basename(file).replace(/.Dockerfile$/, '')
+    return {
+      type: 'docker-image',
+      name,
+      source_file: file,
+      source_directory: path.dirname(file),
+      output_type: outputType,
+      output: sanitizeArtifactName(
+        `${id}__build__output__docker-image__${name}`
+      )
+    }
+  })
+}
+
+async function planBuild(
+  workingDir: string,
+  id: string,
+  imageOutputType: 'file' | 'registry'
+): Promise<BuildPlan[]> {
+  return [
+    ...(await planBuildCharm(workingDir, id)),
+    ...(await planBuildRock(workingDir, id, imageOutputType)),
+    ...(await planBuildDockerImage(workingDir, id, imageOutputType))
```

Comment:
I'm wondering if the planBuild should be an object 
```typescript
{charmBuildPlan: BuildPlan[], rockBuildPlan: BuildPlan[], dockerBuildPlan: BuildPlan[]}
``` 
rather than a list - the reason is because we cannot differentiate different build plans from each other from the output - even though we assume there will be one for each.

Reply:
Rock, charm and docker-image are actually built by the same `build` job, so at here all types of plans are put into a list. This is to simplify the Github action run matrix setup. Thanks!